plutus-core/plutus-core/src/PlutusCore/Builtin/Runtime.hs

```diff
@@ -86,6 +94,28 @@ data BuiltinRuntime val =
                                           -- out what it's going to cost.
         (ToCostingType n)
 
+instance NoThunks (BuiltinRuntime val) where
+    -- Skipping `_denot` in `allNoThunks` check, since it is supposed to be lazy
```

Comment:
From reading the nothunks docs, it's allowed to _force_ expected thunks if we want to check inside them for unexpected thunks. So we could call `wNoThunks` on it, I think.

Although actually... the note on checking for thunks in functions makes me wonder if we need to use the heap-inspection version to look for thunks in closures. I could imagine that e.g. the closures for the costing functions might keep alive a thunk for the cost model representation, which might then not get forced until part way through evaluation!

Reply:
Upshot: think we should call `checkNoThunks` on both `denot` and `costing`.

Reply:
> So we could call wNoThunks on it, I think.

That means although we allow `ToRuntimeDenotationType val n` to be lazy, we don't allow it to contain any other thunks. @effectfully could you confirm if this is the case?

> we need to use the heap-inspection version to look for thunks in closures.

Are you sure we never want to allow thunks in closures? I guess it's fine though to start with a stronger condition, then relax it if it doesn't end up working.

> Upshot: think we should call checkNoThunks on both denot and costing.

What's `checkNoThunks`?

Reply:
> Are you sure we never want to allow thunks in closures?

I think the argument is the same as for not allowing thunks anywhere else in the data structure: it may lead to space leaks, or to computation happening at unexpected times.

If we have 

```
costingFun = \x . ... costModel ....
```
and it doesn't force `costModel`, then `costModel` might not get evaluated until the first time `costingFun` is evaluated, which might be bad.

> What's checkNoThunks

A function which is repeatedly mentioned in the library haddock, but apparently isn't exported :unamused: I think wrapping in the `InspectHeap` newtype might be the only way to do it.

Reply:
In `f = let x = y in z` is `x` considered to be a thunk contained in `f`? We have that for both costing functions and runtime denotations. Although I'm planning to remove the latter. But removing the former would make things less efficient, we do want to have some sharing in there.

Reply:
> In f = let x = y in z is x considered to be a thunk contained in f?

In general yes, unless `x` is inlined and the thunk is somehow optimized away. So sounds like we should just check that the costing functions and denotations are themselves not thunks, then.

Reply:
> I think the argument is the same as for not allowing thunks anywhere else in the data structure: it may lead to space leaks, or to computation happening at unexpected times.
> 
> If we have
> 
> ```
> costingFun = \x . ... costModel ....
> ```
> 
> and it doesn't force `costModel`, then `costModel` might not get evaluated until the first time `costingFun` is evaluated, which might be bad.

I think it's a different story here. We really don't want to create thunks in a loop, but creating thunks when there's a need for sharing is more sensible. And we do have plenty of such thunks. Here's one example:

```haskell
immediateMachineParameters1
  = \ ds_d9e8S ->
      case ds_d9e8S of { CostModel mchnCosts_a9e00 builtinCosts_a9e01 ->
      case runRW#
             (\ s1#_aIS7 ->
                case newArray# 54# arrEleBottom s1#_aIS7 of
                { (# ipv_aISj, ipv1_aISk #) ->
                let { lvl53_sgOXb = paramMkNilPairData builtinCosts_a9e01 } in
                let { lvl54_sgOXa = paramMkNilData builtinCosts_a9e01 } in
                let { lvl55_sgOX9 = paramMkPairData builtinCosts_a9e01 } in
                let { lvl56_sgOX8 = paramSerialiseData builtinCosts_a9e01 } in
                <...>
```

These are thunks contained in a `DefaultMachineParameters` that get forced when we execute appropriate costing functions. They are not useful in any way, but we also have

```haskell
runTwoArgumentModel
  = \ ds_dozv ->
      case ds_dozv of {
        ModelTwoArgumentsConstantCost dt_doY4 ->
          let { c_spW5 = I# dt_doY4 } in (\ _ _ -> c_spW5) `cast` <Co:6>;
```

where `c_spW5` is very useful, we do want to only compute it once and share it afterwards (there are more complicated examples).

Now it would be great if we could force the thunk right after creating it, but these thunks are created by GHC and we have little control over them. In particular, this is the code that generates the thunk:

```haskell
runTwoArgumentModel
    (ModelTwoArgumentsConstantCost c) = lazy $ \_ _ -> c
```

we explicitly use `lazy` here! Since otherwise GHC bluntly disrespects our bottoms and wrecks all the carefully set up sharing. Now we could tell it not to with `-fpedantic-bottoms`, but that is less reliable.

So in theory I'm not aware of any place where forcing thunks immediately after creating them would be inappropriate, but in practice it took me weeks to craft the current alignment in a way that makes GHC compile it reasonably and adding "we don't want to have any thunks whatsoever" to the picture will probably make it infeasible to satisfy such requirements.

Reply:
@zliu41 

> So sounds like we should just check that the costing functions and denotations are themselves not thunks, then.

Except the denotations _are_ thunks currently. I'm going to change that soonish, so we'll need to update the logic here, but currently denotations are thunks and can contain thunks. I'm also planning to reduce the amount of thunks from the internals of the denotations, but realistically a denotation is an arbitrary Haskell function that can depend on arbitrary CAFs, the only way we could force those is by evaluating all paths (including failing ones) of every builtin before creating an `EvalautionContext`, which might actually be worth it, but then that's an entirely separate task.

Reply:
So I think for now the comment should be

```
-- Skipping `_denot` in `allNoThunks` check, since it is supposed to be lazy and is allowed to contain thunks.
```

Reply:
Okay, I'm not too worried about this, but it sounds like it's something we should think about at least. Perhaps we also want to include in the note that we don't believe it will retain any very large data.

I guess here's an example where things could go badly:
- The ledger creates an `EvaluationContext` every time the protocol parameters change, perhaps once every epoch to be safe.
- Perhaps they don't discard the old ones properly?
- Each `EvaluationContext` holds on to the cost model blob forever

That could potentially add up to quite a bit of memory. I think assumption 2 is hopefully false though.

Reply:
> Perhaps they don't discard the old ones properly?

Then it's a space leak that needs to be fixed. Making `EvaluationContext` retain large data will make the space leak more apparent and will allow them to notice and fix it, so it's a feature :)

Reply:
Understandably, they're trying to nail down all the possible space leaks everywhere :) So we're not cooperating with that, but I think I'm happy saying that we don't expect this to retain _much_.